File: addChecksum.py
#!/usr/bin/env python

#############################################################################
# Python script to add a checksum and date/time to an Adblock subscription. #
#                                                                           #
# To add a checksum to a subscription file, run the script like this:       #
#                                                                           #
#   python addChecksum.py subscription.txt                                  #
#                                                                           #
# Note: your subscription file should be saved in UTF-8 encoding, otherwise #
# the generated checksum might be incorrect.                                #
#############################################################################

import sys
import hashlib
import base64
import re
import datetime

def update_modified(contents):
    dt = datetime.datetime.today().strftime("%d %B %Y (%I:%M:%S %p)")
    return re.sub(r"(!\s*Updated\s*:\s*)[^\r\n]*(|(?:\r?\n))", "\\g<1>"+dt+"\\g<2>", contents, 1)

def checksum_data(data):
    # Remove all CR symbols and empty lines.
    ret = data.replace("\r", "")
    ret = re.sub(r"\n+", "\n", ret)

    # Calculate the base64 encoded md5 hash with trailing equal signs removed.
    m = hashlib.md5()
    m.update(ret)
    ret = base64.b64encode(m.digest()).rstrip("=")

    return ret

def usage():
    print(sys.argv[0] + " subscription.txt")

def main(fn):
    # Read the file.
    contents = ""
    with open(fn, "rb") as f:
        contents = f.read()

    # Regular expression used to find the checksum that's embedded in the file.
    rx = re.compile(r"^.*!\s*checksum[\s\-:]+([\w\+\/=]+).*(\r\n|\r|\n)?", re.M|re.I)

    # Get old checksum.
    ochk = rx.search(contents)
    if not ochk:
        return 0 # We don't touch files without existing checksums.
    old_checksum = ochk.group(1)

    # Remove the old checksum.
    contents = rx.sub("", contents)

    # Calculate a new checksum and determine if the file has been modified.
    new_checksum = checksum_data(contents)
    if old_checksum == new_checksum:
        return 0

    # Update the date and time.
    contents = update_modified(contents)

    # Recalculate the checksum as we've altered the date.
    new_checksum = checksum_data(contents)

    # Insert new checksum in file as the second line,
    contents = re.sub(r"(\r\n|\r|\n)", r"\1! Checksum: "+new_checksum+r"\1", contents, 1)

    # Write the new file.
    with open(fn, "w+b") as f:
        f.write(contents)

    print((fn+":").ljust(30) + ("Dated and checksummed".rjust(20)))

    return 0

if __name__=='__main__':
    # Parse command line.
    if len(sys.argv) != 2:
        usage()
        sys.exit(1)

    try:
        sys.exit(main(sys.argv[1]))
    except IOError as e:
        print("I/O error({0}): {1}".format(e.errno, e.strerror))
        sys.exit(1)

File: addChecksum.pl
  binmode(FILE);
  binmode(FILE);
  close(FILE);
  close(FILE);
  exit 0;
  local $/;
  my $file = shift;
  my $result = <FILE>;
  my ($file, $contents) = @_;
  open(local *FILE, "<", $file) || die "Could not read file '$file'";
  open(local *FILE, ">", $file) || die "Could not write file '$file'";
  print FILE $contents;
  return $result;
#
#
#
#                                                                           #
#                                                                           #
#                                                                           #
#                                                                           #
#      http://www.apache.org/licenses/LICENSE-2.0
#   perl addChecksum.pl subscription.txt                                    #
# #!/usr/bin/perl
# Calculate new checksum
# Calculate new checksum: remove all CR symbols and empty
# checksums and date and time of last modification to downloadable          #
# Copyright 2011 Wladimir Palant
# distributed under the License is distributed on an "AS IS" BASIS,
# download and checksum mismatches(broken downloads) will be rejected.      #
# Get existing checksum.
# If the old checksum matches the new one bail.
# Insert checksum into the file
# Licensed under the Apache License, Version 2.0 (the "License");
# limitations under the License.
# lines and get an MD5 checksum of the result (base64-encoded,
# Note: your subscription file should be saved in UTF-8 encoding, otherwise #
# Recalculate the checksum as we've altered the date.
# Remove already existing checksum.
# See the License for the specific language governing permissions and
# subscriptions. The checksum will be validated by Adblock Plus on          #
# the generated checksum might be incorrect.                                #
# This is a slightly modified version of the reference script to add        #
# To add a checksum to a subscription file, run the script like this:       #
# Unless required by applicable law or agreed to in writing, software
# Update the date and time.
# without the trailing = characters).
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#############################################################################
#############################################################################
$checksum = md5_base64($checksumData);
$checksumData = $data;
$checksumData =~ s/\n+/\n/g;
$checksumData =~ s/\n+/\n/g;
$checksumData =~ s/\r//g;
$checksumData =~ s/\r//g;
$data =~ /^.*!\s*checksum[\s\-:]+([\w\+\/=]+).*\n/gmi;
$data =~ s/(!.*Updated:\s*)([^\r\n]*)(\r?\n)/$1$dat$3/;
$data =~ s/(\r?\n)/$1! Checksum: $checksum$1/;
$data =~ s/^.*!\s*checksum[\s\-:]+([\w\+\/=]+).*\n//gmi;
$year += 1900; # Year is years since 1900.
die "Usage: $^X $0 subscription.txt\n" unless @ARGV;
elsif ($hour >= 13) {$hour -= 12; }
if ($checksum eq $oldchecksum)
if ($hour == 0)     {$hour = 12;  }
if ($hour >= 12)    {$ampm = "PM";}
my $ampm = "AM";
my $checksum = md5_base64($checksumData);
my $checksumData = $data;
my $dat = sprintf("$mday $months[$mon] $year (%d:%02d:%02d $ampm)", $hour, $min, $sec);
my $data = readFile($file);
my $file = $ARGV[0];
my $oldchecksum = $1;
my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime();
my @months = qw(Jan Feb Mar Apr May Jun Jul Aug Sep Oct Nov Dec);
sub readFile
sub writeFile
use Digest::MD5 qw(md5_base64);
use strict;
use warnings;
writeFile($file, $data);
{
{
{
}
}
}
